src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -1283,15 +1329,42 @@ void addSubpatternsForTuple(ImmutableArray<TypeWithAnnotations> elementTypes)
             BinaryPatternSyntax node,
             TypeSymbol inputType,
             uint inputValEscape,
+            out TypeSymbol outputType,
+            out uint outputValEscape,
             bool permitDesignations,
             bool hasErrors,
             DiagnosticBag diagnostics)
         {
             bool isDisjunction = node.PatternOperator.Kind() == SyntaxKind.OrKeyword;
-            permitDesignations = permitDesignations & !isDisjunction; // prevent designators under 'or'
-            var left = BindPattern(node.LeftPattern, inputType, inputValEscape, permitDesignations, hasErrors, diagnostics);
-            var right = BindPattern(node.RightPattern, inputType, inputValEscape, permitDesignations, hasErrors, diagnostics);
-            return new BoundBinaryPattern(node, disjunction: isDisjunction, left, right, inputType, hasErrors);
+            if (isDisjunction)
+            {
+                permitDesignations = false; // prevent designators under 'or'
+                var left = BindPattern(node.LeftPattern, inputType, inputValEscape, out var leftOutputType, out var leftOutputValEscape, permitDesignations, hasErrors, diagnostics);
```

Comment:
General question: Why do we care about valEscape in pattern? What's an otherwise non-error scenario where it matters?

Reply:
When an input value (with its corresponding val escape) is captured into a pattern variable, the pattern variable inherits its val escape.  This is a binding-time concept that has no effect on the rest of the compilation so it is dropped after being used to compute the val escape for the pattern variables.  This is just threading that through the binder.

There are no valEscape rules in relational patterns as they cannot define pattern variables.

---
In reply to: [389182670](https://github.com/dotnet/roslyn/pull/42207#discussion_r389182670) [](ancestors = 389182670)